// upscaler/dist/node/shared/src/constants.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidRange = exports.isNumber = exports.hasValidChannels = exports.isValidModelDefinition = exports.ModelDefinitionValidationError = exports.isValidModelType = exports.isString = exports.isTensor = exports.isThreeDimensionalTensor = exports.isFourDimensionalTensor = exports.makeIsNDimensionalTensor = exports.isDynamicShape4D = exports.isFixedShape4D = exports.isShape4D = void 0;
var tf = __importStar(require("@tensorflow/tfjs-core"));
var types_1 = require("./types");
var isShape4D = function (shape) {
    if (!Boolean(shape) || !Array.isArray(shape) || shape.length !== 4) {
        return false;
    }
    return shape.every(function (value) { return value === null || typeof value === 'number'; });
};
exports.isShape4D = isShape4D;
var isFixedShape4D = function (shape) { return (0, exports.isShape4D)(shape) && shape[1] !== null && shape[2] !== null && shape[1] > 0 && shape[2] > 0; };
exports.isFixedShape4D = isFixedShape4D;
var isDynamicShape4D = function (shape) { return (0, exports.isShape4D)(shape) && !(0, exports.isFixedShape4D)(shape); };
exports.isDynamicShape4D = isDynamicShape4D;
function makeIsNDimensionalTensor(rank) {
    function fn(pixels) {
        try {
            return pixels.shape.length === rank;
        }
        catch (err) {
        }
        return false;
    }
    return fn;
}
exports.makeIsNDimensionalTensor = makeIsNDimensionalTensor;
exports.isFourDimensionalTensor = makeIsNDimensionalTensor(4);
exports.isThreeDimensionalTensor = makeIsNDimensionalTensor(3);
var isTensor = function (input) { return input instanceof tf.Tensor; };
exports.isTensor = isTensor;
var isString = function (el) { return typeof el === 'string'; };
exports.isString = isString;
var isValidModelType = function (modelType) { return typeof modelType === 'string' && ['layers', 'graph',].includes(modelType); };
exports.isValidModelType = isValidModelType;
var ModelDefinitionValidationError = (function (_super) {
    __extends(ModelDefinitionValidationError, _super);
    function ModelDefinitionValidationError(type) {
        var _this = _super.call(this, type) || this;
        _this.type = type;
        return _this;
    }
    return ModelDefinitionValidationError;
}(Error));
exports.ModelDefinitionValidationError = ModelDefinitionValidationError;
var isValidModelDefinition = function (modelDefinition) {
    var _a, _b;
    if (modelDefinition === undefined) {
        throw new ModelDefinitionValidationError(types_1.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.UNDEFINED);
    }
    if (!(0, exports.isValidModelType)((_a = modelDefinition.modelType) !== null && _a !== void 0 ? _a : 'layers')) {
        throw new ModelDefinitionValidationError(types_1.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.INVALID_MODEL_TYPE);
    }
    if (!modelDefinition.path && !((_b = modelDefinition._internals) === null || _b === void 0 ? void 0 : _b.path)) {
        throw new ModelDefinitionValidationError(types_1.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.MISSING_PATH);
    }
    return true;
};
exports.isValidModelDefinition = isValidModelDefinition;
var hasValidChannels = function (tensor) { return tensor.shape.slice(-1)[0] === 3; };
exports.hasValidChannels = hasValidChannels;
var isNumber = function (el) { return typeof el === 'number'; };
exports.isNumber = isNumber;
var isValidRange = function (range) { return Array.isArray(range) && range.length === 2 && range.every(exports.isNumber); };
exports.isValidRange = isValidRange;
