// upscaler/dist/browser/esm/shared/src/constants.js
import * as tf from '@tensorflow/tfjs-core';
import { MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE } from './types';
export const isShape4D = (shape) => {
    if (!Boolean(shape) || !Array.isArray(shape) || shape.length !== 4) {
        return false;
    }
    return shape.every((value) => value === null || typeof value === 'number');
};
export const isFixedShape4D = (shape) => isShape4D(shape) && shape[1] !== null && shape[2] !== null && shape[1] > 0 && shape[2] > 0;
export const isDynamicShape4D = (shape) => isShape4D(shape) && !isFixedShape4D(shape);
export function makeIsNDimensionalTensor(rank) {
    function fn(pixels) {
        try {
            return pixels.shape.length === rank;
        }
        catch (err) {
        }
        return false;
    }
    return fn;
}
export const isFourDimensionalTensor = makeIsNDimensionalTensor(4);
export const isThreeDimensionalTensor = makeIsNDimensionalTensor(3);
export const isTensor = (input) => input instanceof tf.Tensor;
export const isString = (el) => typeof el === 'string';
export const isValidModelType = (modelType) => typeof modelType === 'string' && ['layers', 'graph',].includes(modelType);
export class ModelDefinitionValidationError extends Error {
    type;
    constructor(type) {
        super(type);
        this.type = type;
    }
}
export const isValidModelDefinition = (modelDefinition) => {
    if (modelDefinition === undefined) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.UNDEFINED);
    }
    if (!isValidModelType(modelDefinition.modelType ?? 'layers')) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.INVALID_MODEL_TYPE);
    }
    if (!modelDefinition.path && !modelDefinition._internals?.path) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.MISSING_PATH);
    }
    return true;
};
export const hasValidChannels = (tensor) => tensor.shape.slice(-1)[0] === 3;
export const isNumber = (el) => typeof el === 'number';
export const isValidRange = (range) => Array.isArray(range) && range.length === 2 && range.every(isNumber);
